File: magazine/features/material/storage.py
# -*- coding: utf-8 -*-
"""발표 자료 파일 저장.

보안상 지키는 것:
- 저장 파일명은 서버가 만든다. 클라이언트가 준 이름을 경로로 쓰지 않는다.
- 업로드 디렉터리 밖을 가리키는 경로는 거부한다.
- HTML/SVG 는 같은 오리진에서 인라인으로 열면 포털 세션을 노린 XSS 가
  되므로 인라인 허용 목록에서 뺀다.
"""
import mimetypes
import os
import uuid
from typing import Optional
from urllib.parse import quote

from fastapi import HTTPException

from core.config import Settings

INLINE_TYPES = {
    "application/pdf",
    "image/png", "image/jpeg", "image/gif", "image/webp", "image/bmp",
    "text/plain",
}
CHUNK = 1024 * 1024


async def save_upload(settings: Settings, tid: int, upload) -> str:
    """업로드를 저장하고 저장 파일명을 돌려준다. 상한을 넘으면 413."""
    original = os.path.basename(upload.filename or "자료")
    ext = os.path.splitext(original)[1][:16]
    stored = f"{tid}_{uuid.uuid4().hex}{ext}"
    dest = os.path.join(settings.upload_dir, stored)

    size = 0
    try:
        with open(dest, "wb") as out:
            while True:
                chunk = await upload.read(CHUNK)
                if not chunk:
                    break
                size += len(chunk)
                if size > settings.max_upload_bytes:
                    limit = settings.max_upload_bytes // (1024 * 1024)
                    raise HTTPException(status_code=413,
                                        detail=f"{limit}MB 까지 올릴 수 있습니다")
                out.write(chunk)
    except Exception:
        remove(settings, stored)
        raise
    return stored


def remove(settings: Settings, stored: Optional[str]) -> None:
    if not stored:
        return
    try:
        os.remove(os.path.join(settings.upload_dir, stored))
    except OSError:
        pass


def resolve(settings: Settings, stored: Optional[str]) -> str:
    """저장 파일의 실제 경로. 없거나 디렉터리를 벗어나면 404."""
    if not stored:
        raise HTTPException(status_code=404, detail="올라온 파일이 없습니다")
    root = os.path.realpath(settings.upload_dir)
    path = os.path.realpath(os.path.join(settings.upload_dir, stored))
    if os.path.commonpath([path, root]) != root:
        raise HTTPException(status_code=404, detail="올라온 파일이 없습니다")
    if not os.path.exists(path):
        raise HTTPException(status_code=404, detail="파일을 찾을 수 없습니다")
    return path


def disposition(name: str):
    """(media_type, Content-Disposition) — 안전한 타입만 인라인."""
    ctype = mimetypes.guess_type(name)[0] or "application/octet-stream"
    inline = ctype in INLINE_TYPES
    how = "inline" if inline else "attachment"
    return (ctype if inline else "application/octet-stream",
            f"{how}; filename*=UTF-8''{quote(name)}")

File: learn/lib/storage.php
<?php
/**
 * 이수증 파일 저장.
 *
 * 이수증이 발급되지 않는 강의는 강의 사이트의 진행률 화면을 캡쳐한 이미지로 대신한다.
 * 서버는 둘을 구분하지 않는다 — 첨부가 한 장이라도 있으면 청구할 수 있다.
 */

const UPLOAD_DIR    = __DIR__ . '/../var/uploads';
const MAX_UPLOAD_MB = 50;

// HTML/SVG 는 같은 오리진에서 열리면 포털 세션을 노린 XSS 가 된다. 확장자 화이트리스트
// 밖은 아예 받지 않고, 받은 것도 learn_disposition 이 인라인 여부를 다시 판정한다.
const ALLOWED_EXT = ['png', 'jpg', 'jpeg', 'gif', 'webp', 'bmp', 'pdf'];

const INLINE_TYPES = [
    'pdf'  => 'application/pdf',
    'png'  => 'image/png',
    'jpg'  => 'image/jpeg',
    'jpeg' => 'image/jpeg',
    'gif'  => 'image/gif',
    'webp' => 'image/webp',
    'bmp'  => 'image/bmp',
];

function learn_upload_dir() {
    if (!is_dir(UPLOAD_DIR)) @mkdir(UPLOAD_DIR, 0777, true);
    return realpath(UPLOAD_DIR) ?: UPLOAD_DIR;
}

function learn_ensure_allowed($filename) {
    $ext = strtolower(pathinfo((string)$filename, PATHINFO_EXTENSION));
    if (!in_array($ext, ALLOWED_EXT, true)) {
        throw new LearnError('이미지(png, jpg, gif, webp, bmp)와 pdf 만 올릴 수 있습니다', 422);
    }
    return $ext;
}

/** 저장 파일명은 서버가 만든다. 원본명은 DB 에만 둔다. */
function learn_save_upload($rid, array $file) {
    if (($file['error'] ?? UPLOAD_ERR_NO_FILE) !== UPLOAD_ERR_OK) {
        $code = $file['error'] ?? UPLOAD_ERR_NO_FILE;
        if ($code === UPLOAD_ERR_INI_SIZE || $code === UPLOAD_ERR_FORM_SIZE) {
            throw new LearnError(MAX_UPLOAD_MB . 'MB 까지 올릴 수 있습니다', 413);
        }
        throw new LearnError('파일을 올리지 못했습니다', 400);
    }
    if ((int)$file['size'] > MAX_UPLOAD_MB * 1024 * 1024) {
        throw new LearnError(MAX_UPLOAD_MB . 'MB 까지 올릴 수 있습니다', 413);
    }

    $original = basename((string)($file['name'] ?? '이수증'));
    $ext      = learn_ensure_allowed($original);
    $stored   = $rid . '_' . bin2hex(random_bytes(16)) . '.' . $ext;

    if (!move_uploaded_file($file['tmp_name'], learn_upload_dir() . '/' . $stored)) {
        throw new LearnError('파일을 저장하지 못했습니다', 500);
    }
    return [$original, $stored];
}

function learn_remove_upload($stored) {
    if (!$stored) return;
    $path = learn_upload_dir() . '/' . basename((string)$stored);
    if (is_file($path)) @unlink($path);
}

/** 경로 이탈 검사. DB 값이라도 그대로 붙이지 않는다. */
function learn_resolve_upload($stored) {
    if (!$stored) throw new LearnError('올라온 이수증이 없습니다', 404);
    $root = learn_upload_dir();
    $path = realpath($root . '/' . basename((string)$stored));
    if (!$path || strncmp($path, $root, strlen($root)) !== 0 || !is_file($path)) {
        throw new LearnError('파일을 찾을 수 없습니다', 404);
    }
    return $path;
}

/** 화이트리스트 밖은 강제로 내려받기 처리한다 */
function learn_disposition($name) {
    $ext    = strtolower(pathinfo((string)$name, PATHINFO_EXTENSION));
    $inline = isset(INLINE_TYPES[$ext]);
    return [
        $inline ? INLINE_TYPES[$ext] : 'application/octet-stream',
        ($inline ? 'inline' : 'attachment') . "; filename*=UTF-8''" . rawurlencode($name),
    ];
}
